crates/zksync/compiler/src/zksolc/mod.rs

```diff
@@ -86,14 +90,102 @@ pub trait FindContract {
 
     /// Finds a contract matching the ZK bytecode hash
     fn find_zk_bytecode_hash(&self, code_hash: H256) -> Option<&DualCompiledContract>;
+
+    /// Finds a contract matching the ZK deployed bytecode
+    fn find_zk_deployed_bytecode(&self, bytecode: &[u8]) -> Option<&DualCompiledContract>;
+
+    /// Finds a contract own and nested factory deps
+    fn fetch_all_factory_deps(&self, root: &DualCompiledContract) -> HashSet<Vec<u8>>;
 }
 
 impl FindContract for Vec<DualCompiledContract> {
+    fn find_zk_deployed_bytecode(&self, bytecode: &[u8]) -> Option<&DualCompiledContract> {
+        self.iter().find(|contract| bytecode.starts_with(&contract.zk_deployed_bytecode))
+    }
+
     fn find_evm_bytecode(&self, bytecode: &[u8]) -> Option<&DualCompiledContract> {
         self.iter().find(|contract| bytecode.starts_with(&contract.evm_bytecode))
     }
 
     fn find_zk_bytecode_hash(&self, code_hash: H256) -> Option<&DualCompiledContract> {
         self.iter().find(|contract| code_hash == contract.zk_bytecode_hash)
     }
+
+    fn fetch_all_factory_deps(&self, root: &DualCompiledContract) -> HashSet<Vec<u8>> {
```

Comment:
It's probably easier to `self.as_slice().fetch_all_factory_deps()` since it's defined for `&[DualCompiledContract]` below